### PR TITLE
feat(weight-sync): refactor placement and weight reshard

### PR DESCRIPTION
**Pull request by sherlockcooper**
### Description

Refactor the actor→rollout weight synchronization to support:

1. **DPA (Decoupled Per-Attention)** models (e.g. GLM-4.7-Flash) where attention TP, dense MLP TP, and MoE EP use different parallelism — the previous reshard path assumed uniform TP.
2. **Colocate + disaggregated** placement (per-target reshard + DisaggRankMapper rewrite + bucket_length as GPU tensor).
3. **vLLM** backend (new vllm_0_23 worker + dense weight-sync + 0.23 compat).
4. **fc1 unification** — all convertors route fc1 through reshard_fused_fc1.
5. **Inference-side** reshard decoupled from rollout (inference_dst_tp_rank + same-TP skip to avoid redundant all_gather + .contiguous() for P2P send).
6. **Per-tag onload/offload split** — weights restored during sync, KV-cache + cudagraph deferred until after actor releases reshard buffers, avoiding colocate OOM.
7. **Placement–reshard decoupling** — dropped the actor-TP >= rollout-TP constraint; rank mapping uses (engine_id, rank_in_engine) and reshard uses dst_rank/dst_world_size, separating where workers sit (placement) from how weights are sharded (reshard), enabling arbitrary TP ratios.

### Motivation and Context

GLM-4.7-Flash uses DPA (enable_dp_attention), where three layer types (attn / dense MLP / routed expert) have different TP sharding. The previous reshard path assumed uniform TP, breaking colocate weight sync for DPA models. This PR separates the sharding paths, supports arbitrary TP ratios, extends to vLLM + disaggregated mode, and fixes colocate memory (per-tag onload/offload split + deferred KV/cudagraph).

### How has this been tested?

- **GLM-4.7-Flash colocate** (SGLang): model load (MLA+MoE+47L+EP8+MTP) + weight sync (DPA + per-tag onload/offload + deferred KV) + engine init + 1 step + clean exit.
- **Qwen vLLM disagg** (vLLM): engine init (4 rank, vllm_0_23 worker) + HF→Megatron ckpt convert + actor+inference load + weight sync (disagg DisaggRankMapper + per-target reshard + bucket_length GPU tensor) + inference sync (inference_dst_tp_rank + same-TP skip + .contiguous()) + per-tag onload + generation + training backward + clean exit.
- ruff + ruff-format pass.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project. (`ruff` + `ruff-format` pass on all changed files.)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes. (GLM colocate e2e config + CI job)
- [x] All new and existing tests passed. (Manual e2e passed; CI runs via the `run-ci` label.)

### CODE REVIEW

File: requirements/install.sh [hunk 1]
```diff
@@ -3033,11 +3033,11 @@ EOF
...
 
 install_mbridge() {
-    # megatron-bridge 0.4.2 requires python >= 3.12, so RLinf publishes a
+    # megatron-bridge 0.5.0 requires python >= 3.12, so RLinf publishes a
     # py3.10/3.11 fork. --no-deps avoids re-resolving torch and nemo-toolkit.
-    echo "[install.sh] Installing rlinf-megatron-bridge 0.4.2 (PyPI wheel)..."
-    uv pip install --no-deps --extra-index-url https://pypi.org/simple "rlinf-megatron-bridge==0.4.2"
-    echo "[install.sh] rlinf-megatron-bridge 0.4.2 installed (import: megatron.bridge)."
+    echo "[install.sh] Installing rlinf-megatron-bridge 0.5.0 (PyPI wheel)..."
+    uv pip install --no-deps --extra-index-url https://pypi.org/simple "rlinf-megatron-bridge==0.5.0"
```
Comment (andylin-hao) [inline requirements/install.sh]: Could we remove the version pin, so that future version upgrade doesn't require changing the install script. The --extra-index-url is also unnecessary.

File: .github/workflows/agent-e2e-tests.yml [hunk 2]
```diff
@@ -778,3 +778,34 @@ jobs:
...
+        uses: actions/checkout@v5
+
+      - name: Create reason environment
+        run: |
+          unset UV_DEFAULT_INDEX
+          export UV_PATH=/workspace/dataset/.uv
+          export UV_LINK_MODE=symlink
+          export UV_CACHE_DIR=/workspace/dataset/.uv_cache
+          export UV_PYTHON_INSTALL_DIR=/workspace/dataset/.uv_python
+          bash requirements/install.sh agentic --sglang 0.5.12 --torch 2.11.0 --transformers 5.6.0 --no-apex --platform nvidia --use-mirror --mbridge /workspace/dataset/Megatron-Bridge-v0.5.0
```
Comment (andylin-hao) [inline .github/workflows/agent-e2e-tests.yml]: bash requirements/install.sh agentic

**Review by andylin-hao — APPROVED** [review]

File: tests/e2e_tests/reasoning/glm47-flash-grpo-mbridge-sgl.yaml [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+  model:
+    model_type: glm4_moe_lite
+    megatron_checkpoint: null
+    precision: bf16
+    add_bias_linear: False
+
+    tensor_model_parallel_size: 4
+    pipeline_model_parallel_size: 1
+    expert_model_parallel_size: 8
```
Comment (FxxxxU) [inline tests/e2e_tests/reasoning/glm47-flash-grpo-mbridge-sgl.yaml]: Since the CI machine has a limited number of GPUs, we need to scale down the test here.

File: tests/e2e_tests/reasoning/glm47-flash-grpo-mbridge-sgl.yaml [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    experiment_name: "ci-test"
+    logger_backends: ["tensorboard"]
+
+  max_epochs: 1
+  max_steps: 3
+
+  val_check_interval: 1
+  save_interval: 3
+
+  seq_length: 2048
```
Comment (FxxxxU) [inline tests/e2e_tests/reasoning/glm47-flash-grpo-mbridge-sgl.yaml]: A size of 1024 or more can speed up testing.

File: rlinf/hybrid_engines/vllm/vllm_0_23/worker.py [hunk 5]
```diff
@@ -127,9 +132,18 @@ def sync_hf_weight(self) -> None:
             # recv from the Megatron backend
             # Megatron use weight bucket to sync weight, the bucket length in dict of bucket 0, bucket_length
             state_dict.pop("bucket_length")
+            if isinstance(bucket_length, torch.Tensor):
```
Comment (FxxxxU) [inline rlinf/hybrid_engines/vllm/vllm_0_23/worker.py:135]: We need to add an `else` and an `assert` statement here to ensure that any errors on the `actor send` side can be quickly identified.

File: rlinf/utils/convertor/utils.py [hunk 6]
```diff
@@ -733,36 +712,50 @@ def build_rules(self) -> list[ConvertorRule]:
...
                 [r"model.layers.\g<i>.mlp.down_proj.\g<wb>"],
             ),
             # ---- shared expert MLP ----
+            # shared_experts.linear_fc1 is split by tp_gather_fn (unified fc1
+            # branch) into gate_proj/up_proj; the SPLIT_NONE rules below handle
+            # them. linear_fc2 (down_proj) just renames.
             ConvertorRule(
                 re.compile(
-                    rf"decoder\.layers\.{LID}\.mlp\.shared_experts\.linear_fc1\.{WB}$"
+                    rf"decoder\.layers\.{LID}\.mlp\.shared_experts\.linear_fc2\.{WB}$"
```
Comment (FxxxxU) [inline rlinf/utils/convertor/utils.py:738]: I think it would be easier for users to understand if we replaced “linear_fc2” with “down_proj” here, and then when converting ‘transfer’ earlier, the key should be “decoder\.layers\.{LID}\.mlp\.shared_experts\.down_proj.”

File: rlinf/config.py [hunk 7]
```diff
@@ -122,6 +122,9 @@ def __private_create__(cls, value: str) -> "SupportedModel":
 SupportedModel.QWEN3_VL_MOE_SFT = SupportedModel.register("qwen3_vl_moe", force=True)
 SupportedModel.GR00T_N1D6 = SupportedModel.register("gr00t_n1d6", force=True)
 SupportedModel.DEEPSEEK_V3 = SupportedModel.register("deepseek_v3", force=True)
+# GLM-4.7-Flash: MLA (DeepSeek-V3-style) + GLM MoE + MTP, via Megatron-Bridge
+# GLM47FlashBridge (rlinf-megatron-bridge >=0.5.0). Needs mcore 0.18.
```
Comment (FxxxxU) [inline rlinf/config.py]: Is an assert needed to perform a check?

**Review by guozhen1997 — APPROVED** [review]
